Add RAG for Hybrid Search example

aulorbe opened the pull request:
## Problem

Add the notebook that accompanied the RAG for Hybrid Search talk & demo in SF on 10/19. 


## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan

If reviewer(s) can just look through the implementation details to ensure everything was done correctly, that'd be great! Esp the sparse and dense encoding steps for the vectors & the query

jamescalam commented:
Pip installs, should so something like below and include versions (because otherwise your notebook will be broken very soon after langchain/others update their lib)
```
!pip install -qU \
  pinecone-client==2.1.2 \
  unstructured==x.x.x \
  ...
```
I can also see that there are no pip installs in there for langchain need to add that :)

---

Many people will be running these on Google Colab (easiest, least friction method, and also better for those in countries/areas with slower internet speeds, so they're not downloading/uploading tons of vectors etc from all the APIs)

To help those users we add Colab (and nbviewer) shields to our notebooks, you'd just add to top of notebook in a markdown cell with:

```
[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/pinecone-io/examples/blob/master/learn/search/hybrid-search/rag-for-hyrbid/rag-for-hybrid-search.ipynb) [![Open nbviewer](https://raw.githubusercontent.com/pinecone-io/examples/master/assets/nbviewer-shield.svg)](https://nbviewer.org/github/pinecone-io/examples/blob/master/learn/search/hybrid-search/rag-for-hyrbid/rag-for-hybrid-search.ipynb)
```

---

You're in the /search directory, I think this belongs in /generation/rag-for-hybrid — reason being that we try to organize based on use-case (ie from user perspective), so they'd come in looking for RAG and hybrid — ie their use case is generation (that happens to include search, but so do all of our other RAG examples)

aulorbe commented:
Okie dokie @jamescalam , I've now fixed:
- pip installs
- bew installs (although brew wouldn't accept my version commands -- tried both the `==xx` and the `@xx` syntax, but neither worked, so I commented the versions in the cell)
- added the Colab markdown
- moved the nb into the correct dir

LMK if it looks good now!

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<p>Points to local mac path</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

aulorbe replied:
<p>Oh, I left this because users were instructed to download the PDFs locally. I have them in our repo in case users want to download them from us instead of Arxiv.</p><p> </p><p>Do you think they should be read from the repo itself? Since this is just a demo, I was thinking ppl would come w/their own PDFs when trying it out </p>

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<p>is this meant to be commented out?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

aulorbe replied:
<p>It is! As noted in the intro, I've commented out most of the preview cells b/c they're just sooo long in Github when rendered.</p>

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #11.       dimension = 384,  # dimensionality of our vectors
</pre><p>might want to clarify it is of the dense vectors</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='74'/>

aulorbe replied:
<p>Oh, nice! Yes I'll do that</p>

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<p>In what case would this fail</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='82'/>

aulorbe replied:
<p>If there were more IDs than chunked files. Is that what you mean?</p>

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #1.    query_sembedding = bm25.encode_documents(query)
</pre><p> </p><p>bm25.encode_query</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='96'/>

aulorbe replied:
<p>Oh, shoot, thank you.</p>

gdj0nes commented on learn/generation/rag-for-hybrid/rag-for-hybrid-search.ipynb:
```diff
@@ -0,0 +1,2569 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #3.    def weight_by_alpha(sparse_embedding: Dict[str, List[Any]], dense_embedding: List[float], alpha: float) -&gt; Tuple[Dict[str, List[Any]], List[float]]:
</pre><p>You can import this from pinecone-text but understand if you are are trying to be more instructive</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pinecone-io/examples/pull/284/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='101'/>

aulorbe replied:
<p>Yeah I was trying to be a bit verbose here :)</p>

jamescalam commented:
Looks good, one tweak to make it easier for users is to download the PDFs directly from this repo, will work in colab too (as shown in screenshot):
<img width="1450" alt="image" src="https://github.com/pinecone-io/examples/assets/35938317/ba35b9df-2ac5-448a-8fc7-f622db1f3587">

So:

```python
def get_pdf(url):
    res = requests.get(url)
    filename = url.split("/")[-1]
    with open(filename, 'wb') as f:
        f.write(res.content)

github_dir = "https://github.com/pinecone-io/examples/blob/audrey_hybrid/learn/generation/rag-for-hybrid/"
files = ["freshdiskann_paper.pdf", "hnsw_paper.pdf", "ivfpq_paper.pdf"]

for f in files:
    get_pdf(github_dir+f)
```

Just remember to swap `audrey_hybrid` in the URL above to `master` before merging

aulorbe commented:
> Looks good, one tweak to make it easier for users is to download the PDFs directly from this repo, will work in colab too (as shown in screenshot): <img alt="image" width="1450" src="https://user-images.githubusercontent.com/35938317/279803300-ba35b9df-2ac5-448a-8fc7-f622db1f3587.png">
> 
> So:
> 
> ```python
> def get_pdf(url):
>     res = requests.get(url)
>     filename = url.split("/")[-1]
>     with open(filename, 'wb') as f:
>         f.write(res.content)
> 
> github_dir = "https://github.com/pinecone-io/examples/blob/audrey_hybrid/learn/generation/rag-for-hybrid/"
> files = ["freshdiskann_paper.pdf", "hnsw_paper.pdf", "ivfpq_paper.pdf"]
> 
> for f in files:
>     get_pdf(github_dir+f)
> ```
> 
> Just remember to swap `audrey_hybrid` in the URL above to `master` before merging

Oo I didn't know about this. Thanks!
